File: nameMap/nameMap/ZwqClass/view/popView.m
//
//  popView.m
//  StaffMap
//
//  Created by zhangwenqiang on 2017/3/14.
//  Copyright © 2017年 张文强. All rights reserved.
//

#import "popView.h"
@interface popView()
@property (retain,nonatomic)NSMutableArray<UIImageView*>*SubItems;
@property (retain,nonatomic)UIView*BottomView;
@end
@implementation popView
-(void)setItems:(NSArray*)pAry
{
    for (NSString*pImgName in pAry) {
        UIImageView*pImgView=[[UIImageView alloc]initWithFrame:CGRectMake(0, 0, 100, 100)];
        pImgView.image=[UIImage imageNamed:pImgName];
        UITapGestureRecognizer*pTap=[[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(onTap:)];
        [pImgView addGestureRecognizer:pTap];
        pImgView.userInteractionEnabled=YES;
        [self.SubItems addObject:pImgView];
    }
}
-(NSMutableArray*)SubItems
{
    if (!_SubItems) {
        _SubItems=[[NSMutableArray alloc]initWithCapacity:10];
    }
    return _SubItems;
}
-(void)popViews
{
    if (!_BottomView||!_BottomView.superview)
    {
        UIView*pScreenView=[UIApplication sharedApplication].keyWindow.rootViewController.view;
        _BottomView=[[UIView alloc]initWithFrame:pScreenView.bounds];
        UITapGestureRecognizer*pTap=[[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(onTap:)];
        [_BottomView addGestureRecognizer:pTap];
        [pScreenView addSubview:_BottomView];
        CGRect rtPop=[self.superview convertRect:self.frame toView:pScreenView];
        CGPoint ptDir=[self getPopOrientation:rtPop Boundto:pScreenView.bounds];
        for (UIImageView*pImgView in _SubItems) {
            [_BottomView addSubview:pImgView];
            NSUInteger i=[_SubItems indexOfObject:pImgView]+1;
            CGAffineTransform t=CGAffineTransformMakeTranslation(ptDir.x*100*i,ptDir.y*100*i);
            pImgView.center=CGPointApplyAffineTransform(CGPointMake(CGRectGetMidX(rtPop), CGRectGetMidY(rtPop)),t) ;
        }
    }
}
-(void)shrinkViews
{
    [_BottomView removeFromSuperview];
}
-(CGPoint)getPopOrientation:(CGRect)rect Boundto:(CGRect)rtBound
{
    int x=1;
    int y=1;
    int a=rect.origin.x;
    int b=rect.origin.y;
    int c=rtBound.size.width-CGRectGetMaxX(rect);
    int d=rtBound.size.height-CGRectGetMaxY(rect);
    if( a>c) {
        x = -1;
    }
    if (b>d ){
        y = -1;
    }
    if (MAX(a, c)>MAX(b,d)) {
        // x*=2
        y=0;
    }else{
        //y*=2
        x=0;
    }
    return CGPointMake(x, y);
}
-(void) onTap:(UITapGestureRecognizer*)pTab{
    [self shrinkViews];
    UIView* pView=pTab.view;
    if ([pView isKindOfClass:[UIImageView class]]) {
        NSInteger nIndex=[_SubItems indexOfObject:(UIImageView*)pView];
        if(nIndex!=NSNotFound)
        {
            [self handleSubitemTap:nIndex];
        }
        
    }
    
}
-(void) handleSubitemTap:(NSUInteger)nIndex{
    //self
    if (_delegate) {
        [_delegate handleSubitemDidTap:nIndex];
    }
    
    
}

@end

File: nameMap/nameMap/ZwqClass/view/popView.swift
//
//  popView.swift
//  StaffMap
//
//  Created by zhangwenqiang on 2017/3/13.
//  Copyright © 2017年 张文强. All rights reserved.
//

import UIKit
class popView: UIImageView {
    var subItems: [UIImageView]=Array.init();
    var ButView:UIView?=nil;
    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */
    func setItems(items:Array<String>)->Bool
    {
        for item in items {
            let SubView:UIImageView=UIImageView.init(frame:CGRect.init(x: 0, y: 0, width: 100, height: 100))
            SubView.image=UIImage.init(contentsOfFile: item)
            subItems.append(SubView)
            
        }
        return true;
    }
    func popViews() -> Bool {
        let ScreenView=UIApplication.shared.keyWindow?.rootViewController?.view
        ButView=UIView.init(frame:(ScreenView?.bounds)!)
        ButView?.addGestureRecognizer(UITapGestureRecognizer.init(target: self, action:Selector(("onTap:"))))
        ScreenView?.addSubview(ButView!)
        let rtPop:CGRect=(self.superview?.convert(self.frame, to:ScreenView))!
        let ptDir:CGPoint=self.getPopOrientation(rect:rtPop,Bound:(ScreenView?.bounds)!)
        for item in self.subItems {
            ButView?.addSubview(item)
//            if abs(ptDir.x)==2 {
//                ptDir.y = 0
//            }
//            else
//            {
//                ptDir.x=0
//            }
            let i:CGFloat=CGFloat(subItems.index(of: item)!)
            let t:CGAffineTransform=CGAffineTransform.init(translationX: ptDir.x*self.frame.size.width*i, y: ptDir.y*self.frame.size.height*i)
            item.center.applying(t)
        }
        return true;
    }
    func shrinkViews() -> Void {
        ButView?.removeFromSuperview()
    }
    func getPopOrientation(rect:CGRect,Bound:CGRect) -> CGPoint {
        var x:Int=1
        var y:Int=1
        let ScreenView=UIApplication.shared.keyWindow?.rootViewController?.view
        
        let a=self.frame.origin.x;
        let b=self.frame.origin.y;
        let c=(ScreenView?.frame.size.width)!-rect.maxX
        let d=(ScreenView?.frame.size.height)!-rect.maxY
        if a>c {
            x = -1
        }
        if b>d {
            y = -1
        }
        if max(a, c)>max(b,d) {
           // x*=2
            y=0
        }else{
            //y*=2
            x=0
        }
        return CGPoint.init(x: x, y: y)
    }
    func onTap(pTap:UIGestureRecognizer) -> Void {
        self.shrinkViews();
        self.handleTap();
    }
    func handleTap() -> Void {
        
    }
}
